src/main/groovy/com/github/jengelman/gradle/plugins/shadow/ShadowExtension.groovy

```diff
@@ -1,33 +1,36 @@
 package com.github.jengelman.gradle.plugins.shadow
 
 import org.gradle.api.Project
+import org.gradle.api.Task
 import org.gradle.api.artifacts.ProjectDependency
 import org.gradle.api.artifacts.SelfResolvingDependency
+import org.gradle.api.provider.Provider
 import org.gradle.api.publish.maven.MavenPom
 import org.gradle.api.publish.maven.MavenPublication
-import org.gradle.util.GradleVersion
 
 class ShadowExtension {
+    private final Provider<Task> shadowJar
+    private final Provider<List<Dep>> allDependencies
 
-    void component(Project project, MavenPublication publication) {
-
-        if (GradleVersion.current() >= GradleVersion.version("6.6")) {
-            publication.artifact(project.tasks.named("shadowJar"))
-        } else {
-            publication.artifact(project.tasks.shadowJar)
-        }
-
-        final def allDependencies = project.provider {
-            project.configurations.shadow.allDependencies.collect {
-                if ((it instanceof ProjectDependency) || ! (it instanceof SelfResolvingDependency)) {
-                    new Dep(it.group, it.name, it.version)
-                }
+    ShadowExtension(Project project) {
+        shadowJar = project.provider { project.tasks.getByName("shadowJar") }
+        allDependencies = project.provider {
+            project.configurations.getByName("shadow").allDependencies.findAll {
+                it instanceof ProjectDependency || it instanceof SelfResolvingDependency
+            }.collect {
+                new Dep(it.group, it.name, it.version)
             }
         }
+    }
+
+    void component(MavenPublication publication) {
+        publication.artifact(shadowJar)
+
         publication.pom { MavenPom pom ->
             pom.withXml { xml ->
                 def dependenciesNode = xml.asNode().get('dependencies') ?: xml.asNode().appendNode('dependencies')
-                allDependencies.get().each {
+                final def allDeps = allDependencies
```

Comment:
I meant to have this local variable outside `pom.withXml {}` closure

Reply:
Tried both before `publication.pom` or before `pom.withXml` or before `allDependencies.get()`, both failed.

Reply:
Seems `publication.artifact` jsut need `archiveFile` as the input instead of `ShadowJar`, that's why:
```
- Task `:c:generatePomFileForShadowPublication` of type `org.gradle.api.publish.maven.tasks.GenerateMavenPom`: cannot serialize object of type 'com.github.jengelman.gradle.plugins.shadow.tasks.ShadowJar', a subtype of 'org.gradle.api.Task', as these are not supported with the configuration cache.
```
fixed it in 69ffd437498353e3e5a8b957062595a9f6d6b448

Reply:
Good to merge?